File: android-wife-assistant/app/src/test/java/com/wifeassistant/ReviewTest.kt
package com.wifeassistant

import com.wifeassistant.data.Feedback
import com.wifeassistant.data.Review
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

// اختبار حساب التقرير (Review.compute) — نظير review.test.js.
class ReviewTest {
    private fun fb(slot: String, choice: String, themes: List<String> = listOf("امتنان", "دعاء")) =
        Feedback("2026-07-05", slot, themes, choice, null)

    @Test
    fun `نسبة القبول بتتحسب وبتتجاهل regen`() {
        val report = Review.compute(
            listOf(
                fb("morning", "pick1"),
                fb("morning", "ignore"),
                fb("evening", "regen"),
            ),
            styleCount = 5,
        )
        assertEquals(2, report.total) // regen مش محسوب
        assertEquals(1, report.accepted)
        assertEquals(50, report.acceptRate)
        assertEquals(5, report.styleExamplesCount)
    }

    @Test
    fun `أعلى المواضيع بتظهر`() {
        val report = Review.compute(
            listOf(
                fb("morning", "pick1", listOf("امتنان", "دعاء")),
                fb("evening", "pick2", listOf("دعم", "امتنان")),
            ),
            styleCount = 0,
        )
        assertTrue(report.topThemes.any { it.first == "امتنان" })
    }

    @Test
    fun `خانة بتتجاهل بنسبة عالية بتترصد`() {
        val report = Review.compute(
            List(4) { fb("evening", "ignore") },
            styleCount = 0,
        )
        assertNotNull(report.worstSlot)
        assertEquals("evening", report.worstSlot!!.slot)
    }

    @Test
    fun `تقرير فاضي مبيقعش`() {
        val report = Review.compute(emptyList(), styleCount = 0)
        assertEquals(0, report.total)
        assertEquals(0, report.acceptRate)
        assertNull(report.worstSlot)
    }
}

File: telegram-wife-assistant/test/review.test.js
const { test, beforeEach, after } = require('node:test');
const assert = require('node:assert');
const { useTempStore } = require('./helpers');

const tmp = useTempStore();
const store = require('../store');
const review = require('../review');

beforeEach(() => tmp.cleanup());
after(() => tmp.cleanup());

test('نسبة القبول بتتحسب صح وبتتجاهل regen', () => {
  store.addFeedback({ slot: 'morning', themesShown: ['امتنان', 'دعاء'], choice: 'pick1' });
  store.addFeedback({ slot: 'morning', themesShown: ['تقدير', 'دعم'], choice: 'ignore' });
  store.addFeedback({ slot: 'evening', themesShown: ['اشتياق', 'دعاء'], choice: 'regen' });
  const { data } = review.buildReport();
  // regen مش بيتحسب، فالقرارات = 2، المقبول = 1 → 50%
  assert.strictEqual(data.total, 2);
  assert.strictEqual(data.accepted, 1);
  assert.strictEqual(data.acceptRate, 50);
});

test('أعلى المواضيع نجاحاً بتظهر', () => {
  store.addFeedback({ slot: 'morning', themesShown: ['امتنان', 'دعاء'], choice: 'pick1' });
  store.addFeedback({ slot: 'evening', themesShown: ['دعم', 'امتنان'], choice: 'pick2' });
  const { data } = review.buildReport();
  const topLabels = data.topThemes.map((t) => t[0]);
  assert.ok(topLabels.includes('امتنان'));
});

test('اقتراح إيقاف خانة بتتجاهل بنسبة عالية', () => {
  for (let i = 0; i < 4; i++) {
    store.addFeedback({ slot: 'evening', themesShown: ['دعم', 'دعاء'], choice: 'ignore' });
  }
  const { text, data } = review.buildReport();
  assert.strictEqual(data.worstSlot.slot, 'evening');
  assert.match(text, /توقفها أو تغيّر ميعادها/);
});

test('تقرير فاضي مبيقعش (0 من 0)', () => {
  const { data } = review.buildReport();
  assert.strictEqual(data.total, 0);
  assert.strictEqual(data.acceptRate, 0);
});
